Docs: Clarified Migration README: grammar, out-of-date, organization (#2141)
bottlerocket-os/bottlerocket: sources/api/migration/README.md

## Overview

We're not perfect, so we need the ability to fix problems in our design, but it's critical to maintain user configuration through those changes.
To achieve both, the ability to migrate configuration between versions is required.

The mechanism is reminiscent of database migrations - individual bits of code that manipulate specific data, and that are labeled to indicate what version they apply to.

## When to migrate data

Migrations run during the boot process when it is detected that the on-disk data store doesn't match the version of the booting image.
The system will download any missing, appropriate migrations before rebooting.

One downside of this approach is that migration failure requires another reboot to flip back to a matching version.

> In return, this provides a consistent process: always check and migrate the data store before it's used. Additionally, pre-reboot process is simpler.

Reviewer: ```suggestion
In return, this provides a consistent process: always check and migrate the data store before it's used.
Additionally, the pre-reboot process is simpler.
```

The process can also handle some cases of unexpected version mismatches at boot, for example if something goes wrong during an update.

### Offline migration

Since migrations run at boot the network connection may not be available yet.
This means it is impossible to check for fixes to migrations or download missing migrations.

> When downloading an OS update, the TUF metadata is retrieved for that particular variant and architecture. This metadata describes the migrations necessary between various updates.

Reviewer: Sentences on newlines is intentional since it produces better diffs but still renders on the same line. Also this was factually incorrect since manifest.json is not TUF metadata:

```suggestion
When downloading an OS update, the TUF metadata and the Bottlerocket update manifest are retrieved for that particular variant and architecture.
The manifest describes the migrations necessary between various updates and the TUF metadata ensures its integrity.
```

## Integration with update system

> As mentioned above in [Offline Migration](#offline-migration), the updated metadata lists the migrations needed to move between versions.

Reviewer 2: ```suggestion
As mentioned above in [Offline Migration](#offline-migration), the update metadata lists the migrations needed to move between versions.
```

> As mentioned above in [Offline Migration](#offline-migration), the updated metadata lists the migrations needed to move between versions.

Reviewer: ```suggestion
As mentioned above in [Offline Migration](#offline-migration), the manifest lists the migrations needed to move between versions.
```

> This metadata is made available for the migrator to use later to verify the migrations.

Reviewer: ```suggestion
This manifest is made available for the migrator later along with TUF metadata to verify the migrations.
```

There may be times when we discover issues with migrations after release and need to replace them.

> These fixed migrations are listed in the updated metadata and downloaded along with the image.

Reviewer 2: ```suggestion
These fixed migrations are listed in the update metadata and downloaded along with the image.
```

> These fixed migrations are listed in the updated metadata and downloaded along with the image.

Reviewer: ```suggestion
These fixed migrations are listed in the manifest and downloaded along with the image.
```

Since this metadata is the source of truth for the migration list, any old migrations will simply be ignored.

## How to update and flip the data store

To start, the migration system copies the data store, updates it by running migrations, and (when ready) flips symlinks to the new version.

> This duplicates the data but it isn't expected to be large quantities of data in the data store.

Reviewer 2: This is pretty awkward, maybe:
```suggestion
This duplicates the data, but currently the data store only contains a small amount of data.
```

## Integration with update system

> As mentioned above in [Offline Migration](#offline-migration), the mainfest lists the migrations needed to move between versions.

Reviewer 2: ```suggestion
As mentioned above in [Offline Migration](#offline-migration), the manifest lists the migrations needed to move between versions.
```

This manifest is made available for the migrator later along with TUF metadata to verify the migrations.

There may be times when we discover issues with migrations after release and need to replace them.

> These fixed migrations are listed in the mainfest and downloaded along with the image.

Reviewer: ```suggestion
These fixed migrations are listed in the manifest and downloaded along with the image.
```

   -> v1.5.2
   -> v1.5.2_0123456789abcdef
```

> The migration system appends a random identifier to a the final level so it can track migration attempts and help prevent timing issues.

Reviewer 2: ```suggestion
The migration system appends a random identifier to the final level so it can track migration attempts and help prevent timing issues.
```

Old versions can be kept for quick rollbacks, but automated cleanup will be necessary to prevent filling the disk.

> Note that the version applies to both the `live` and `pending` trees, which both live inside the directory described above. Both live and pending data is migrated to prevent user configuration information from being lost.

Reviewer: Nit: consistency with the new-sentence on a newline pattern.

```suggestion
Note that the version applies to both the `live` and `pending` trees, which both live inside the directory described above.
Both live and pending data is migrated to prevent user configuration information from being lost.
```

Author: FWIW, the newline pattern is not universal to all markdown parsers. If we build a website and we want to include files with this pattern they may render is individual paragraphs. 

In this use case, I get it though.

Migrations are Rust code adhering to a Migration interface.
This is defined in the `migration-helpers` library; see [Helpers](#helpers).

The interface requires handling forward and backward migrations through specific methods.
Each method will give data and metadata maps as input, and require data and metadata maps as output.
(These structures are nested, rather than using dotted keys, to make it easier to handle sub-trees and reduce dependency on existing data store code.)

Migration code should not assume that any given keys exist, because migrations will be run on live data (where all keys will likely exist) and on pending data (where none, some, or all keys may exist).
Plus, different variants of Bottlerocket may not have the same keys.

> The migration system deserializes the function outputs into the incoming model types to confirm that the structure is valid.

Reviewer 2: This was more of a TODO comment; my sense is that it's not implemented.

Author: Ah. Gotcha. Inline it's not super clear that it's a todo. 

What if we change the **Open questions** heading to **Open questions and future directions** and move this down to that section? That way all the maybe stuff goes into end of the document.

Author: Changed in [73d2b39](https://github.com/bottlerocket-os/bottlerocket/pull/2141/commits/73d2b39eebdc48fa6dd7d77bae97207342828693)

To write a migration, start a Rust project at `/migrations/<applicable version>/migrate-<name>/Cargo.toml`

> The filename of the migration is used by the migration system to order migrations.

Reviewer: I think we changed this behavior to preserve the manifest order:

```suggestion
Migrations run in the order that they are found in the manifest.
```

The name takes the format `migrate_v<applicable version>_<name>`.
Cargo does not allow naming binaries this way, so the migration build process renames them appropriately when installing them into the image.

### Helpers

> There is a standard structure for migration code that handles common things like argument parsing, this allows for a common CLI interface for the migration system.

Reviewer 2: ```suggestion
There is a standard structure for migration code that handles common things like argument parsing, which allows for a common CLI interface for the migration system.
```
